asserts/validation_set.go

```diff
@@ -0,0 +1,262 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// Presence represents a presence constraint.
+type Presence string
+
+var (
+	PresenceRequired Presence = "required"
+	PresenceOptional Presence = "optional"
+	PresenceInvalid  Presence = "invalid"
+)
+
+func presencesAsStrings(presences ...Presence) []string {
+	strs := make([]string, len(presences))
+	for i, pres := range presences {
+		strs[i] = string(pres)
+	}
+	return strs
+}
+
+var validValidationSetSnapPresences = presencesAsStrings(PresenceRequired, PresenceOptional, PresenceInvalid)
+
+func checkPresence(snap map[string]interface{}, which string, valid []string) (Presence, error) {
+	presence, err := checkOptionalStringWhat(snap, "presence", which)
+	if err != nil {
+		return Presence(""), err
+	}
+	if presence != "" && !strutil.ListContains(valid, presence) {
+		return Presence(""), fmt.Errorf("presence %s must be one of %s", which, strings.Join(valid, "|"))
+	}
+	return Presence(presence), nil
+}
+
+// ValidationSetSnap holds the details about a snap constrained by a validation-set assertion.
+type ValidationSetSnap struct {
+	Name   string
+	SnapID string
+
+	Presence Presence
+
+	Revision int
+}
+
+// SnapName implements naming.SnapRef.
+func (s *ValidationSetSnap) SnapName() string {
+	return s.Name
+}
+
+// ID implements naming.SnapRef.
+func (s *ValidationSetSnap) ID() string {
+	return s.SnapID
+}
+
+func checkValidationSetSnap(snap map[string]interface{}) (*ValidationSetSnap, error) {
+	name, err := checkNotEmptyStringWhat(snap, "name", "of snap")
+	if err != nil {
+		return nil, err
+	}
+	if err := naming.ValidateSnap(name); err != nil {
+		return nil, fmt.Errorf("invalid snap name %q", name)
+	}
+
+	what := fmt.Sprintf("of snap %q", name)
+
+	snapID, err := checkStringMatchesWhat(snap, "id", what, naming.ValidSnapID)
+	if err != nil {
+		return nil, err
+	}
+
+	presence, err := checkPresence(snap, what, validValidationSetSnapPresences)
+	if err != nil {
+		return nil, err
+	}
+
+	var snapRevision int
+	if _, ok := snap["revision"]; ok {
+		var err error
+		snapRevision, err = checkSnapRevisionWhat(snap, "revision", what)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if snapRevision != 0 && presence == PresenceInvalid {
+		return nil, fmt.Errorf(`cannot specify revision %s at the same time as stating its presence is invalid`, what)
```

Comment:
What do you think a more compact: "cannot specify revision %s and state its presence as invalid" ?

Reply:
I think the longer version is a bit clearer